README.md
# TheDeadBurn_v17

Factorio mod that gives you burnable monster bodies when you kill biters.
Can be used as fuel.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 2020-04-20
  Features:
  Changes:
    - Updated for Factorio 1.1
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.18.2
Date: 2020-04-20
  Features:
    - Added some biter bodies to crude oil recipes
    - Added an option to move "Coal Liquefaction" under "Chemical Science"
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.18.1
Date: 2020-01-22
  Features:
  Changes:
  Bugfixes:
    - Fixed icon size
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.18.0
Date: 2020-01-22
  Features:
  Changes:
    - Updated for 0.18
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 2020-01-22
  Features:
  Changes:
  Bugfixes:
    - Fixed item icon scaling on the ground + on belts
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 2020-01-22
  Features:
  Changes:
  Bugfixes:
    - Icon sizes needed tweaking
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2020-01-22
  Features:
  Changes:
  Bugfixes:
  Info:
    - Updated for 0.18
---------------------------------------------------------------------------------------------------
Version: 0.1.8
Date: 2020-01-14
  Features:
  Changes:
    - Biter bodies recipes updated; icons, names and descriptions
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.6
Date: 2020-01-12
  Features:
  Changes:
  Bugfixes:
    - Icon fixes so they display correctly in ALT mode
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.5
Date: 2020-01-12
  Features:
    - Added recipe to turn monster bodies into heavy oil, for kickstarting coal liquefaction
  Changes:
    - Changed fuel values for monster bodies to something more reasonable
    - Added thumbnail
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.4
Date: 2019-10-18
  Features:
  Changes:
  Bugfixes:
    - Accidentally overwrite global coal recipe, fixed.
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.3
Date: 2019-09-28
  Features:
    - Biter fuel now is size-dependent, and can be fed through a smelter to
      turn into coal. Now you can run your whole factory on biters.
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 2019-09-28
  Features:
  Changes:
  Bugfixes:
    - Trouble fixing locale
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2019-09-28
  Features:
    - Drops different fuel sizes depending on biter size/spawner
  Changes:
  Bugfixes:
    - Fixed typos, etc
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-11-04
  Features:
  Changes:
  Bugfixes:
  Info:
    - Initial Release of Mod, fixed for 0.17 from
      https://mods.factorio.com/mod/TheDeadBurn


data.lua
require("prototypes.item")
require("prototypes.recipe")

local function starts_with(str, start)
    return str:sub(1, #start) == start
end

local function AddLootToEntity(entityType, entityName, itemName, probability, countMin, countMax)
    if data.raw[entityType] ~= nil then
        if data.raw[entityType][entityName] ~= nil then
            if data.raw[entityType][entityName].loot == nil then
                data.raw[entityType][entityName].loot = {}
            end
            table.insert(data.raw[entityType][entityName].loot, {item = itemName, probability = probability, count_min = countMin, count_max = countMax})
        end
    end
end

for _, unit in pairs(data.raw["unit"]) do
    if unit.corpse then
        if starts_with(unit.corpse, "small") then
            AddLootToEntity("unit", unit.name, "monster-body-small", 1, 1, 1)
        end
        if starts_with(unit.corpse, "medium") then
            AddLootToEntity("unit", unit.name, "monster-body-medium", 1, 1, 1)
        end
        if starts_with(unit.corpse, "big") then
            AddLootToEntity("unit", unit.name, "monster-body-big", 1, 1, 1)
        end
        if starts_with(unit.corpse, "behemoth") then
            AddLootToEntity("unit", unit.name, "monster-body-behemoth", 1, 1, 1)
        end
    end
end

for _, unit in pairs(data.raw["unit-spawner"]) do
    if unit.corpse then
        AddLootToEntity("unit-spawner", unit.name, "monster-body-medium", 1, 1, 1)
    end
end

for _, unit in pairs(data.raw["turret"]) do
    if unit.corpse then
        if starts_with(unit.corpse, "small") then
            AddLootToEntity("turret", unit.name, "monster-body-small", 1, 1, 1)
        end
        if starts_with(unit.corpse, "medium") then
            AddLootToEntity("turret", unit.name, "monster-body-medium", 1, 1, 1)
        end
        if starts_with(unit.corpse, "big") then
            AddLootToEntity("turret", unit.name, "monster-body-big", 1, 1, 1)
        end
        if starts_with(unit.corpse, "behemoth") then
            AddLootToEntity("turret", unit.name, "monster-body-behemoth", 1, 1, 1)
        end
    end
end

if settings.startup["early_coal_liqification"].value then
    local tech = data.raw.technology["coal-liquefaction"]
    tech.prerequisites = {"advanced-oil-processing"}
    tech.unit = {
        count = 200,
        ingredients = {
            {"automation-science-pack", 1},
            {"logistic-science-pack", 1},
            {"chemical-science-pack", 1}
        },
        time = 30
    }
end


prototypes/item.lua
local function starts_with(str, start)
    return str:sub(1, #start) == start
end

for _, corpse in pairs(data.raw["corpse"]) do
    if corpse.subgroup == "corpses" then
        local tint
        local monsterBody = table.deepcopy(data.raw["item"]["coal"])
        monsterBody.name = "monster-body"
        monsterBody.icon = "__base__/graphics/icons/medium-biter-corpse.png"
        monsterBody.icon_size = 64

        if starts_with(corpse.name, "small") then
            monsterBody.name = "monster-body-small"
            monsterBody.fuel_value = "4MJ"
        elseif starts_with(corpse.name, "medium") then
            monsterBody.name = "monster-body-medium"
            tint = {r = 255, g = 0, b = 0}
            monsterBody.fuel_value = "32MJ"
        elseif starts_with(corpse.name, "big") then
            monsterBody.name = "monster-body-big"
            tint = {r = 0, g = 0, b = 255}
            monsterBody.fuel_value = "64MJ"
        elseif starts_with(corpse.name, "behemoth") then
            monsterBody.name = "monster-body-behemoth"
            tint = {r = 0, g = 255, b = 0}
            monsterBody.fuel_value = "128MJ"
            monsterBody.fuel_acceleration_multiplier = 1.5
            monsterBody.fuel_top_speed_multiplier = 1.5
        else
            monsterBody.name = "monster-body-small"
            monsterBody.fuel_value = "4MJ"
        end

        monsterBody.icon_mipmaps = 4
        monsterBody.icons = {{icon = monsterBody.icon, icon_size = 64, tint = tint}}
        monsterBody.pictures = {{filename = monsterBody.icon, tint = tint, mipmap_count = 4, scale = 0.25, size = monsterBody.icon_size}}

        monsterBody.icon = nil
        monsterBody.dark_background_icon = nil

        data:extend({monsterBody})
    end
end


prototypes/recipe.lua
Constants = require("constants")

data:extend(
    {
        {
            group = "intermediate-products",
            name = "biter-liquefaction",
            order = "c",
            type = "item-subgroup"
        }
    }
)

local function add_smelter_recipe(ingredient, ing_qty, res, res_qty, recipe_energy, order)
    local recipe = {
        type = "recipe",
        name = ingredient,
        category = "smelting",
        energy_required = recipe_energy,
        ingredients = {{ingredient, ing_qty}},
        products = {{res, res_qty}},
        order = "biter-bodies-" .. order,
        main_product = "",
        subgroup = "raw-resource",
        result = res,
        result_count = res_qty
    }
    local coal_item = data.raw.item["coal"]
    recipe.icons = {{icon = coal_item.icon, icon_size = coal_item.icon_size}}
    for _, icon_layer in pairs(data.raw.item[ingredient].icons) do
        table.insert(recipe.icons, icon_layer)
    end

    data:extend({recipe})
end

local function add_oil_recipe(ingredient, ing_qty, recipe_energy, order, tint)
    local body_oil_recipe = {
        name = ingredient .. "-biter-liquefaction",
        allow_decomposition = false,
        category = "crafting-with-fluid",
        enabled = false,
        energy_required = recipe_energy,
        icon_mipmaps = 4,
        icon_size = 64,
        icons = {
            {icon = "__base__/graphics/icons/fluid/barreling/barrel-fill.png", icon_mipmaps = 4, icon_size = 64},
            {icon = "__base__/graphics/icons/fluid/barreling/barrel-fill-side-mask.png", icon_mipmaps = 4, icon_size = 64, tint = {a = 0.75, b = 0, g = 0, r = 0}},
            {icon = "__base__/graphics/icons/fluid/barreling/barrel-fill-top-mask.png", icon_mipmaps = 4, icon_size = 64, tint = {a = 0.5, b = 0.5, g = 0.5, r = 0.5}},
            {icon = "__base__/graphics/icons/fluid/crude-oil.png", icon_mipmaps = 4, icon_size = 64, scale = 0.25, shift = {4, -8}},
            {icon = "__base__/graphics/icons/medium-biter-corpse.png", tint = tint, icon_mipmaps = 4, icon_size = 64, scale = 0.25, shift = {-4, -8}}
        },
        ingredients = {
            {amount = 50, catalyst_amount = 50, name = "steam", type = "fluid"},
            {amount = ing_qty, name = "monster-body-" .. ingredient, type = "item"},
            {amount = 1, catalyst_amount = 1, name = "empty-barrel", type = "item"}
        },
        order = "c[fill-crude-oil-barrel]" .. order,
        results = {
            {amount = 1, catalyst_amount = 1, name = "crude-oil-barrel", type = "item"}
        },
        subgroup = "biter-liquefaction",
        type = "recipe"
    }

    data:extend({body_oil_recipe})

    local tech = data.raw.technology["fluid-handling"]
    table.insert(tech.effects, {recipe = body_oil_recipe.name, type = "unlock-recipe"})
end

add_smelter_recipe("monster-body-small", 1, "coal", 1, 1, "a")
add_smelter_recipe("monster-body-medium", 1, "coal", 8, 8, "b")
add_smelter_recipe("monster-body-big", 1, "coal", 16, 16, "c")
add_smelter_recipe("monster-body-behemoth", 1, "coal", 32, 32, "d")

add_oil_recipe("small", 125, 10, "a", nil)
add_oil_recipe("medium", 25, 10, "b", {r = 255, g = 0, b = 0})
add_oil_recipe("big", 5, 10, "c", {r = 0, g = 0, b = 255})
add_oil_recipe("behemoth", 1, 10, "d", {r = 0, g = 255, b = 0})
